File: python/SherlockValidString.py
class Solution(object):
  def print_yes_or_no(self, char_array):
    char_counts = self.get_char_counts(char_array)
    
    count_counts = self.get_count_counts(char_counts)
    
    print_yes = self.del_one_or_fewer(count_counts)
    
    if print_yes:
      print 'YES'
    else:
      print 'NO'
  
  def get_char_counts(self, char_array):
    char_counts = {}
    
    for c in char_array:
      char_count = char_counts.get(c, 0)
      char_counts.update({c: char_count + 1})
    
    return char_counts
  
  def get_count_counts(self, char_counts):
    count_counts = {}
    
    for char, char_count in char_counts.items():
      char_count_count = count_counts.get(char_count, 0)
      count_counts.update({char_count: char_count_count + 1})
    
    return count_counts
  
  def del_one_or_fewer(self, count_counts):
    if len(count_counts) < 2:
      return True
    elif len(count_counts) == 2:
      at_least_one_one = False
      
      for k, v in count_counts.items():
        if v == 1:
          at_least_one_one = True
        
        if at_least_one_one:
          one_one_in_set = False
          min_count = 2 ^ 32
          max_count = 0
          
          for k, v in count_counts.items():
            if (k == 1 and v == 1):
              one_one_in_set = True
            
            if k < min_count:
              min_count = k
            
            if k > max_count:
              max_count = k
            
          if (one_one_in_set or ((max_count - 1 == min_count) and count_counts.get(max_count) == 1)):
            return True
          
    return False


solution = Solution()

input_string = raw_input()

solution.print_yes_or_no(input_string)

File: java/SherlockValidString.java
import java.io.*;
import java.util.*;
import java.util.concurrent.*;
import java.text.*;
import java.math.*;
import java.util.regex.*;

public class Solution {
    
    private static void printYesNo(char[] charArray) {
        Map<Character, Long> charCounts = getCharCounts(charArray);
        
        Map<Long, Long> countCounts = getCountCounts(charCounts);
        
        boolean printYes = delOneOrFewer(countCounts);
        
        if (printYes) {
            System.out.println("YES");
        } else {
            System.out.println("NO");
        }
    }
    
    // Count characters
    private static Map<Character, Long> getCharCounts(char[] charArray) {
        Map<Character, Long> charCounts = new ConcurrentHashMap<>();
        
        for (int i = 0; i < charArray.length; i++) {
            Long charCount = charCounts.get(charArray[i]);
            charCount = charCount == null ? 0 : charCount;
            charCounts.put(charArray[i], charCount + 1);
        }
        
        return charCounts;
    }
    
    // Count character counts
    private static Map<Long, Long> getCountCounts(Map<Character, Long> charCounts) {
        Map<Long, Long> countCounts = new ConcurrentHashMap<>();
        
        for (Map.Entry<Character, Long> entry : charCounts.entrySet()) {
            long charCount = entry.getValue();
            Long charCountCount = countCounts.get(charCount);
            charCountCount = charCountCount == null ? 0 : charCountCount;
            
            countCounts.put(charCount, charCountCount + 1);
        }
        
        return countCounts;
    }
    
    private static boolean delOneOrFewer(Map<Long, Long> countCounts) {
        if (countCounts.size() < 2) {
            return true;
        } else if (countCounts.size() == 2) {
            // At least one value should be one
            boolean atLeastOneOne = false;
            for (long v : countCounts.values()) {
                if (v == 1) {
                    atLeastOneOne = true;
                }
            }
            
            if (atLeastOneOne) {
                boolean OneOneInSet = false;
                long minCount = Long.MAX_VALUE;
                long maxCount = 0;
                
                for (Map.Entry<Long, Long> entry : countCounts.entrySet()) {
                    if(entry.getKey() == 1 && entry.getValue() == 1) {
                        OneOneInSet = true;
                    }
                    
                    if(entry.getKey() < minCount) {
                        minCount = entry.getKey();
                    }
                    if(entry.getKey() > maxCount) {
                        maxCount = entry.getKey();
                    }
                }
                
                if (OneOneInSet || ((maxCount - 1 == minCount) && countCounts.get(maxCount) == 1)) {
                    return true;
                }
            }
        }
        
        return false;
    }

    public static void main(String[] args) {
        /* Enter your code here. Read input from STDIN. Print output to STDOUT. Your class should be named Solution. */
        Scanner scan = new Scanner(System.in);
        String inputString = scan.nextLine();
        char[] inputCharArray = inputString.toCharArray();
        printYesNo(inputCharArray);
    }
}